File: internal/core/engine.go
package core

import (
	"fmt"
	"os"
)

// Engine is the shared core that all runx tools use.
// It orchestrates ref resolution, caching, trust, and materialization.
type Engine struct {
	Flags *Flags
	cache *Cache
	git   *Git
	trust *TrustStore
}

// NewEngine constructs an Engine from flags.
func NewEngine(f *Flags) *Engine {
	return &Engine{
		Flags: f,
		cache: NewCache(f.CacheDir),
		git:   &Git{Verbose: f.Verbose},
		trust: NewTrustStore(NewCache(f.CacheDir).TrustFile()),
	}
}

// Prepare resolves ref, ensures the repo is cached, checks trust, materializes
// a checkout, and returns a ready Workspace.
//
// If mutable is true, the workspace is a writable temporary copy; the caller
// must call ws.Close() when done. Immutable workspaces need not be closed.
func (e *Engine) Prepare(ref RepoRef, mutable bool) (*Workspace, error) {
	// Override ref with --commit flag.
	if e.Flags.Commit != "" {
		ref = makeRepoRef(ref.CanonicalURL, e.Flags.Commit)
	}

	if err := e.cache.EnsureRoots(); err != nil {
		return nil, fmt.Errorf("cache setup: %w", err)
	}

	bareDir := e.cache.BareRepoDir(ref.CanonicalURL)

	// Clone or fetch.
	if err := e.git.CloneOrFetch(
		ref.CanonicalURL, bareDir,
		e.Flags.Offline, e.Flags.Refresh,
		ref.IsMutableRef(), ref.Ref,
	); err != nil {
		return nil, err
	}

	// Resolve ref → commit.
	commit, err := e.git.ResolveRef(bareDir, ref.Ref)
	if err != nil {
		return nil, err
	}

	// Announce the resolved commit (always visible; helps with reproducibility).
	short := commit
	if len(short) > 12 {
		short = short[:12]
	}
	fmt.Fprintf(os.Stderr, "runx: %s → %s\n", ref.DisplayRef(), short)

	// Print pinned command when requested.
	if e.Flags.Pin && ref.IsMutableRef() {
		fmt.Fprintf(os.Stderr, "runx: pin → %s@%s\n", ref.CanonicalURL, commit)
	}

	// Trust check (may prompt the user interactively).
	if err := e.trust.Check(ref, commit, e.Flags); err != nil {
		return nil, err
	}

	// Materialize immutable tree if not already cached.
	treeDir := e.cache.TreeDir(ref.CanonicalURL, commit)
	if !e.cache.TreeExists(ref.CanonicalURL, commit) {
		if e.Flags.Offline {
			return nil, fmt.Errorf("commit %s is not cached and --offline is set", short)
		}
		if err := e.git.Archive(bareDir, commit, treeDir); err != nil {
			// Clean up partial tree on failure.
			_ = os.RemoveAll(treeDir)
			return nil, fmt.Errorf("materializing tree: %w", err)
		}
	}

	if !mutable {
		return ImmutableWorkspace(treeDir, commit), nil
	}
	return MutableWorkspace(treeDir, commit)
}

File: src/core/engine.rs
use crate::core::{Cache, Flags, Git, RepoRef, RunxError, TrustStore, Workspace};
use std::fs;

/// The shared core that all runx tools use.
/// Orchestrates ref resolution, caching, trust, and materialization.
pub struct Engine {
    pub flags: Flags,
    cache: Cache,
    git: Git,
    trust: TrustStore,
}

impl Engine {
    pub fn new(flags: Flags) -> Self {
        let cache = Cache::new(flags.cache_dir.clone());
        let trust_path = cache.trust_file();
        Self {
            git: Git {
                verbose: flags.verbose,
            },
            trust: TrustStore::new(trust_path),
            cache,
            flags,
        }
    }

    /// Resolve ref, ensure repo is cached, check trust, materialize a checkout,
    /// and return a ready Workspace.
    ///
    /// If mutable is true, the workspace is a writable temporary copy.
    pub fn prepare(&self, repo_ref: &RepoRef, mutable: bool) -> Result<Workspace, RunxError> {
        // Override ref with --commit flag.
        let repo_ref = if let Some(ref commit) = self.flags.commit {
            RepoRef {
                canonical_url: repo_ref.canonical_url.clone(),
                clone_url: repo_ref.clone_url.clone(),
                ref_name: commit.clone(),
                is_commit: crate::core::repo::is_full_sha(commit),
            }
        } else {
            repo_ref.clone()
        };

        self.cache.ensure_roots()?;

        let bare_dir = self.cache.bare_repo_dir(&repo_ref.canonical_url);

        // Clone or fetch using the original clone URL (preserves SSH, file://, etc).
        self.git.clone_or_fetch(
            &repo_ref.clone_url,
            &bare_dir,
            self.flags.offline,
            self.flags.refresh,
            repo_ref.is_mutable_ref(),
            &repo_ref.ref_name,
        )?;

        // Resolve ref → commit.
        let commit = self.git.resolve_ref(&bare_dir, &repo_ref.ref_name)?;

        // Announce the resolved commit.
        let short = if commit.len() > 12 {
            &commit[..12]
        } else {
            &commit
        };
        eprintln!("runx: {} \u{2192} {}", repo_ref.display_ref(), short);

        // Print pinned command when requested.
        if self.flags.pin && repo_ref.is_mutable_ref() {
            eprintln!("runx: pin \u{2192} {}@{}", repo_ref.canonical_url, commit);
        }

        // Trust check (may prompt the user interactively).
        self.trust.check(&repo_ref, &commit, &self.flags)?;

        // Materialize immutable tree if not already cached.
        // git archive operates on the local bare repo — no network needed.
        // Archive into a temp dir then rename atomically to prevent TOCTOU
        // races when multiple runx processes target the same commit.
        let tree_dir = self.cache.tree_dir(&repo_ref.canonical_url, &commit);
        if !self.cache.tree_exists(&repo_ref.canonical_url, &commit) {
            let tmp_dir = tree_dir.with_extension(format!("tmp.{}", std::process::id()));
            if let Err(e) = self.git.archive(&bare_dir, &commit, &tmp_dir) {
                let _ = fs::remove_dir_all(&tmp_dir);
                return Err(RunxError::Git {
                    op: "materializing tree",
                    detail: e.to_string(),
                });
            }
            // Atomic move into place; if another process won the race, discard ours.
            if fs::rename(&tmp_dir, &tree_dir).is_err() {
                let _ = fs::remove_dir_all(&tmp_dir);
            }
        }

        if !mutable {
            return Ok(Workspace::immutable(tree_dir, commit));
        }
        Workspace::mutable_copy(&tree_dir, commit)
    }
}
